docs(deploy): local launch templates for v0.16-alpha (PR 30a) (#4483)
QwenLM/qwen-code: docs/users/qwen-serve-deploy-local.md

# Local launch templates for `qwen serve` (v0.16-alpha)

Reference templates for running `qwen serve` as a long-lived background process on a developer workstation. Pairs with the [v0.16-alpha known limits](./qwen-serve.md#v016-alpha-known-limits) — local-only, single-user, BYO bearer token. Containerized / multi-host / TLS-fronted deployments defer to v0.16.x.

> **Audience**: dogfooding developers who want the daemon up across reboots, with logs going somewhere durable, and a clean `restart-on-failure` story. If you only need the daemon for the duration of a single shell session, plain `qwen serve` (foreground, Ctrl-C to stop) is fine.

## Generate a bearer token (once)

```bash
openssl rand -hex 32 > ~/.qwen-serve-token  # user-managed, NOT a built-in path
chmod 600 ~/.qwen-serve-token
export QWEN_SERVER_TOKEN="$(cat ~/.qwen-serve-token)"
```

The path / filename is yours to choose; v0.16-alpha does not auto-generate or auto-locate a token file (deferred to v0.16.x). See the [Authentication](./qwen-serve.md#authentication) section of the user guide for the canonical BYO setup.

The same `QWEN_SERVER_TOKEN` env var is honored by:

- The daemon's `--token` CLI flag (read at boot)

> - The TypeScript SDK's `DaemonClient` constructor (PR 27 fallback — clients with `export QWEN_SERVER_TOKEN=...` in their shell never need to thread the value through their script)

Reviewer: **[Suggestion]** The bullet "The daemon's `--token` CLI flag (read at boot)" reads as if the env var is consumed by the CLI flag, but they're two independent input paths the daemon reads from. The list structure implies the bullets below are consumers of the env var, but the first item is actually a parallel mechanism.

Suggested rewrite of lines 18-24:

```suggestion
The daemon reads the bearer token from either `--token <value>` on the CLI or the `QWEN_SERVER_TOKEN` env var (whitespace stripped from both). The TypeScript SDK's `DaemonClient` constructor falls back to `QWEN_SERVER_TOKEN` when no `token` option is passed (PR 27 fallback — clients with the env var set never need to thread the value through their script).

One shell-level `export` covers both server boot and SDK client construction.
```

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — replaced the muddled bullet list with two paragraphs that read straight: "The daemon reads the bearer token from either `--token <value>` on the CLI or the `QWEN_SERVER_TOKEN` env var (whitespace stripped from both). The TypeScript SDK's `DaemonClient` constructor falls back to `QWEN_SERVER_TOKEN` when no `token` option is passed (PR 27 fallback...). One shell-level `export` covers both server boot and SDK client construction."

So one shell-level `export` covers both server and client.

## Linux: systemd user unit

`~/.config/systemd/user/qwen-serve.service`:

```ini
[Unit]
Description=Qwen Code daemon (loopback HTTP + SSE)
After=network.target

[Service]
Type=simple
# Replace with your project; %h expands to $HOME under user units.
WorkingDirectory=%h/your-project

> ExecStart=/usr/local/bin/qwen serve --bind 127.0.0.1

Author: Adopted in 54d8111fc — source-verified at `packages/cli/src/commands/serve.ts:58`: the CLI flag is `--hostname` (with `--port`), not `--bind`. All 4 templates (systemd / launchd / nohup / tmux) had the wrong flag and would have failed at startup with "unknown option". Switched to `--hostname 127.0.0.1 --port 4170`. Defaults are 127.0.0.1:4170 already, but explicit is better for copy-paste docs. Thanks for catching this!

# DO NOT COMMIT this file with a real token. Use a sealed-secret
# / chezmoi / sops / git-crypt setup if you check the unit file in.

> Environment=QWEN_SERVER_TOKEN=PASTE-YOUR-TOKEN-HERE

Reviewer: **[Suggestion]** The systemd unit embeds the token inline via `Environment=`, but the user already created a `chmod 600` token file in the setup step. With inline `Environment=`, the token lives in the unit file (typically `644`) — readable by any local process or user, negating the `chmod 600` protection. The same issue applies to the system-wide alternative at `/etc/systemd/system/` (world-readable by every local user).

Use `EnvironmentFile=` to reference the existing token file:

```suggestion
EnvironmentFile=%h/.qwen-serve-token
```

For this to work, restructure the token file to contain `QWEN_SERVER_TOKEN=<value>` (not just the raw value):

```bash
echo "QWEN_SERVER_TOKEN=$(cat ~/.qwen-serve-token)" > ~/.qwen-serve-token-env
chmod 600 ~/.qwen-serve-token-env
```

Then use `EnvironmentFile=%h/.qwen-serve-token-env`. Keep the "DO NOT COMMIT" comment.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — switched the systemd unit from inline `Environment=QWEN_SERVER_TOKEN=...` to `EnvironmentFile=%h/.qwen-serve-token-env`. Added a setup step that wraps the existing chmod-600 token file in `KEY=value` form for systemd to read. The unit file (typically 644) no longer contains the token. Same change applies to the system-wide alternative; called that out explicitly in the trailing paragraph.

Restart=on-failure
RestartSec=5
StandardOutput=journal
StandardError=journal

[Install]
WantedBy=default.target
```

Manage:

```bash
systemctl --user daemon-reload

> systemctl --user enable --now qwen-serve.service

Reviewer: **[Critical]** The systemd user-unit section omits `loginctl enable-linger`, which is mandatory for the stated goal of "the daemon up across reboots." Without linger, the user's systemd instance only runs while at least one login session is active. On a headless dev box (SSH in → set up → SSH out), the user manager shuts down on logout and takes the service with it; on reboot the daemon doesn't start until the user logs in again.

```suggestion
systemctl --user daemon-reload
systemctl --user enable --now qwen-serve.service
loginctl enable-linger "$(whoami)"                # keep user manager running after logout
journalctl --user -u qwen-serve -f          # tail logs
systemctl --user restart qwen-serve.service # after token rotation
systemctl --user disable --now qwen-serve.service
```

Add a sentence explaining that without `enable-linger`, the user-level systemd instance only runs while the user is logged in.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — accurate Critical catch. Without `loginctl enable-linger`, the user-level systemd instance shuts down at logout and only starts on next login — completely defeating the "across reboots" goal of the doc on a headless dev box. Added `loginctl enable-linger "$(whoami)"` to the manage block + a paragraph explaining why it's required.

journalctl --user -u qwen-serve -f          # tail logs
systemctl --user restart qwen-serve.service # after token rotation
systemctl --user disable --now qwen-serve.service
```

**System-wide alternative** (shared dev hosts, less common): drop the unit at `/etc/systemd/system/qwen-serve@.service` with `User=%i`, manage via `sudo systemctl enable --now qwen-serve@<username>.service`. Same `[Service]` body otherwise. Pick user-level for single-user workstations.

## macOS: launchd user agent

`~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>com.qwenlm.qwen-serve</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/qwen</string>
    <string>serve</string>
    <string>--bind</string>
    <string>127.0.0.1</string>

>   </array>

Author: Adopted in 54d8111fc — same `--bind` → `--hostname` fix applied to the launchd `ProgramArguments` array (also added explicit `--port 4170` for clarity).

  <!-- launchd does NOT expand `~` or `$HOME` — use absolute paths. -->
  <key>WorkingDirectory</key>
  <string>/Users/YOUR-USERNAME/your-project</string>
  <key>EnvironmentVariables</key>
  <dict>
    <!-- DO NOT COMMIT this file with a real token. -->
    <key>QWEN_SERVER_TOKEN</key>

>     <string>PASTE-YOUR-TOKEN-HERE</string>

Reviewer: **[Suggestion]** The plist file embeds the bearer token inline in `EnvironmentVariables`, but the doc doesn't advise setting restrictive permissions on the plist file itself. By default, files in `~/Library/LaunchAgents/` may have `644` permissions, making the token readable by any local user.

Add after the plist block:

```bash
chmod 600 ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist
```

This mirrors the `chmod 600` guidance already given for the token file in the setup section.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — added `chmod 600 ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist` to the manage block. Plist embeds the inline token; without chmod 600 the token is world-readable in the default 644 perms. Also extended the inline XML comment near the token line to call this out.

  </dict>
  <key>RunAtLoad</key>
  <true/>
  <key>KeepAlive</key>

>   <true/>

Reviewer: **[Suggestion]** `KeepAlive=<true/>` restarts on **all** exits including clean SIGTERM (`exit 0`). This differs from the systemd template's `Restart=on-failure`, which only restarts on non-zero exits. A macOS operator who sends `kill <pid>` expecting a clean stop will see the daemon immediately respawn — the only way to stop it is `launchctl unload`. The "Restart and crash behavior" section says both "work as expected" without calling out this semantic difference.

To match systemd's restart-on-failure behavior:

```suggestion
  <key>KeepAlive</key>
  <dict>
    <key>SuccessfulExit</key>
    <false/>
  </dict>
  <key>ThrottleInterval</key>
  <integer>10</integer>
```

`ThrottleInterval` also prevents restart storms on persistent failures (the systemd template has `RestartSec=5` for the same purpose).

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — bare `<true/>` replaced with `<dict><key>SuccessfulExit</key><false/></dict>` to match systemd `Restart=on-failure` semantics. Added `ThrottleInterval=10` to mirror systemd's `RestartSec=5` and prevent restart storms on persistent failures. Also expanded the "Restart and crash behavior" section to call out the systemd vs launchd semantic difference + ThrottleInterval rationale.

  <key>StandardOutPath</key>
  <string>/tmp/qwen-serve.out.log</string>

>   <key>StandardErrorPath</key>

Reviewer: **[Suggestion]** Logging to `/tmp/` has two problems on macOS:

1. **Security**: `/tmp/` is world-writable (sticky bit). On shared workstations, predictable paths like `/tmp/qwen-serve.err.log` are vulnerable to symlink attacks — a local user can pre-create the file as a symlink to redirect daemon writes.
2. **Log destruction on reload**: launchd truncates `StandardOutPath`/`StandardErrorPath` on every `load`. The token-rotation instructions (`unload` → `load`) silently destroy all prior diagnostic logs — exactly when you need them most.
3. **Periodic cleanup**: macOS `periodic-daily` cleans `/tmp` files older than 3 days.

Use a user-scoped log directory:

```suggestion
  <string>/Users/YOUR-USERNAME/Library/Logs/qwen-serve/out.log</string>
  <key>StandardErrorPath</key>
  <string>/Users/YOUR-USERNAME/Library/Logs/qwen-serve/err.log</string>
```

Add `mkdir -p ~/Library/Logs/qwen-serve` before `launchctl load`. Consider noting that `newsyslog(8)` or a cron job can handle rotation.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — switched both `StandardOutPath` and `StandardErrorPath` to `/Users/YOUR-USERNAME/Library/Logs/qwen-serve/{out,err}.log`. Added `mkdir -p ~/Library/Logs/qwen-serve` to the manage block + an explicit paragraph noting that launchd truncates these on every `load`, so the unload→load token-rotation cycle wipes prior diagnostic logs (back them up if investigating an incident before rotating). Three concerns (symlink attack on /tmp, truncate-on-load, periodic-daily cleanup) all addressed.

  <string>/tmp/qwen-serve.err.log</string>
</dict>
</plist>
```

Manage:

```bash
launchctl load   ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist
launchctl unload ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist  # to stop
tail -f /tmp/qwen-serve.out.log /tmp/qwen-serve.err.log
```

After editing the plist (e.g., rotating the token) you must `unload` then `load` again — `launchctl` does not auto-reload on plist changes the way `systemd daemon-reload` does.

## tmux session (interactive supervision)

> ```bash

Reviewer: **[Suggestion]** The tmux template doesn't pass `QWEN_SERVER_TOKEN` or `--hostname 127.0.0.1` (same `--bind` → `--hostname` issue as Copilot flagged). The systemd and launchd templates both configure both token and bind address; the nohup template configures bind address. This template's omission is inconsistent and could lead users who generate a token above to run an unauthenticated daemon.

```suggestion
tmux new -d -s qwen-serve "cd ~/your-project && QWEN_SERVER_TOKEN=\$QWEN_SERVER_TOKEN qwen serve --hostname 127.0.0.1"
```

Or add a note: "Assumes `QWEN_SERVER_TOKEN` is exported in your shell (see above)."

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — fixed `--bind` → `--hostname 127.0.0.1` (overlap with copilot's findings). Added explicit pre-section note: "Assumes `QWEN_SERVER_TOKEN` is already exported in your shell" + a follow-up paragraph noting that `tmux new -d` inherits the parent shell's environment, so the export flows through automatically.

tmux new -d -s qwen-serve "cd ~/your-project && qwen serve"
tmux attach -t qwen-serve   # see live logs; Ctrl-b d to detach
tmux kill-session -t qwen-serve
```

Best when you want to occasionally watch the daemon's stdout (auth warnings, MCP discovery progress, slow-client warnings) without committing to a service unit. Survives terminal close but not host reboot.

## nohup one-liner (quick + dirty)

```bash

> nohup qwen serve --bind 127.0.0.1 > qwen-serve.log 2>&1 &

Reviewer: **[Suggestion]** The nohup template is missing workspace binding — unlike the tmux template above (which does `cd ~/your-project &&`), this command doesn't set the working directory. The daemon defaults to `process.cwd()`, so running from `~` or `/tmp` will silently bind an unintended workspace, causing `POST /session` calls with the expected `cwd` to return `400 workspace_mismatch`.

```suggestion
nohup bash -c 'cd ~/your-project && qwen serve --bind 127.0.0.1' > qwen-serve.log 2>&1 &
```

Or add `--workspace ~/your-project` to the command.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — wrapped the nohup command in `bash -c 'cd ~/your-project && qwen serve --hostname 127.0.0.1'` so the daemon binds the right workspace regardless of where the user runs the command. Added a paragraph explaining the foot-gun: without `cd`, daemon defaults to `process.cwd()` and `POST /session` from a client expecting the project workspace would silently return `400 workspace_mismatch`.

> nohup qwen serve --bind 127.0.0.1 > qwen-serve.log 2>&1 &

Author: Adopted in 54d8111fc — `--bind` → `--hostname` in the nohup template too. (Folded with the matching workspace-cd fix from a different reviewer thread; nohup template is now `bash -c 'cd ~/your-project && qwen serve --hostname 127.0.0.1'`.)

echo $!  # daemon PID; capture if you want to `kill` cleanly later
```

OK for one-off "let me run this in the background while I poke at the API" workflows. **Not recommended** for anything beyond a single session — no restart-on-crash, log file grows unbounded, no clean way to find the daemon if you forget the PID. Prefer tmux for interactive supervision or systemd / launchd for anything you want to outlast a reboot.

## Verifying the daemon is up

```bash
curl http://127.0.0.1:4170/health                                   # → {"status":"ok"}
curl -H "Authorization: Bearer $QWEN_SERVER_TOKEN" \
  http://127.0.0.1:4170/capabilities | jq .protocolVersions         # daemon's feature set
```

> `/health` is exempted from auth on loopback binds; `/capabilities` always requires auth. If `/capabilities` returns 401, the unit / plist token doesn't match the env-exported token your `curl` is using.

Author: Adopted in 54d8111fc — accurate catch. Reworded the "Verifying the daemon is up" section to clarify that auth is only required when a token is configured (or `--require-auth` is set), with a follow-up note that the templates above all configure a token so the `Authorization` header is needed in practice for the curl examples. Removed the false "always requires auth" claim.

## Token rotation

1. Generate a new token: `openssl rand -hex 32 > ~/.qwen-serve-token-new`
2. Edit the unit file / plist / shell export with the new value
3. Restart the daemon:
   - **systemd**: `systemctl --user restart qwen-serve.service`
   - **launchd**: `launchctl unload ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist && launchctl load ~/Library/LaunchAgents/com.qwenlm.qwen-serve.plist`

>    - **tmux / nohup**: `kill <pid>` then re-run with the new token in env

Author: Adopted in 54d8111fc — token rotation step 1 now writes both the raw `~/.qwen-serve-token` AND the systemd `~/.qwen-serve-token-env` (KEY=value form for `EnvironmentFile=`), and chmod 600's both. Mirrors the initial generation block.

4. Update any client SDKs / scripts. The TypeScript SDK's `DaemonClient` reads `QWEN_SERVER_TOKEN` automatically (PR 27 fallback) — re-`export` the new value in any client shell and reconstruct the client.

## Restart and crash behavior

Service-manager restart works as expected (systemd `Restart=on-failure`, launchd `KeepAlive=true`). Sessions are in-memory and re-attach via SSE `Last-Event-ID` resume per the [Durability model](./qwen-serve.md#durability-model) section of the user guide. Cross-restart durability of session content (prompts, tool calls, conversation history) is **NOT** in v0.16-alpha — a daemon restart drops sessions; clients reconnect and start fresh.

> ## Out of scope (defers to v0.16.x or later)

Reviewer: **[Suggestion]** This paragraph is self-contradictory: it says sessions "re-attach via SSE `Last-Event-ID` resume" then immediately says "a daemon restart drops sessions." SSE `Last-Event-ID` resume only works for client disconnects within a **single daemon process lifetime** — the replay ring is in-memory and does not survive restarts.

Suggested rewrite:

```suggestion
Service-manager restart behavior differs: systemd `Restart=on-failure` skips clean exits; launchd `KeepAlive` restarts on all exits (see the launchd section for `SuccessfulExit` tuning). Client disconnects from a **running** daemon recover via SSE `Last-Event-ID` resume per the [Durability model](./qwen-serve.md#durability-model). A daemon **restart** drops all in-memory sessions — clients reconnect and start fresh. Cross-restart durability of session content (prompts, tool calls, conversation history) is **NOT** in v0.16-alpha.
```

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 54d8111fc — rewrote the "Restart and crash behavior" section to clearly distinguish WITHIN-process client disconnects (covered by SSE Last-Event-ID resume against the in-memory ring) from RESTART events (drops everything; cross-restart durability not in v0.16-alpha). Also documented the per-launcher restart semantics (systemd `Restart=on-failure` vs launchd `KeepAlive` with `SuccessfulExit=false` vs tmux/nohup no-restart).

# Local launch templates for `qwen serve` (v0.16-alpha)

Reference templates for running `qwen serve` as a long-lived background process on a developer workstation. Pairs with the [v0.16-alpha known limits](./qwen-serve.md#v016-alpha-known-limits) — local-only, single-user, BYO bearer token. Containerized / multi-host / TLS-fronted deployments defer to v0.16.x.

> **Audience**: dogfooding developers who want the daemon up across reboots, with logs going somewhere durable, and a clean `restart-on-failure` story. If you only need the daemon for the duration of a single shell session, plain `qwen serve` (foreground, Ctrl-C to stop) is fine.

## Generate a bearer token (once)

```bash
openssl rand -hex 32 > ~/.qwen-serve-token  # user-managed, NOT a built-in path
chmod 600 ~/.qwen-serve-token

> export QWEN_SERVER_TOKEN="$(cat ~/.qwen-serve-token)"

Reviewer: **[Suggestion]** The setup section encourages `export QWEN_SERVER_TOKEN="$(cat ~/.qwen-serve-token)"` as a shell-wide export without warning against adding it to `.bashrc` / `.zshrc`. On a shared workstation, a profile-level export exposes the bearer token to every process spawned from that shell (IDE subprocesses, browser debuggers, npm scripts from unrelated projects). The systemd `EnvironmentFile=` and launchd plist `EnvironmentVariables` already scope the token to just the daemon process — consider recommending those for persistent setups:

> **Note:** This `export` is scoped to the current shell session. Avoid adding it to your shell profile (`.bashrc` / `.zshrc`) on shared workstations — use the systemd `EnvironmentFile=` or launchd plist `EnvironmentVariables` instead so the token is only visible to the daemon process.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 265891f23 — added a callout block right after the `export QWEN_SERVER_TOKEN=...` setup step warning against adding it to `.bashrc` / `.zshrc` on shared workstations (profile-level export exposes the token to every child process — IDE subprocesses, browser debuggers, `npm` scripts from unrelated projects). Points users at the systemd `EnvironmentFile=` / launchd `EnvironmentVariables` mechanisms below for persistent setups since both scope the token to just the daemon process.

```

The path / filename is yours to choose; v0.16-alpha does not auto-generate or auto-locate a token file (deferred to v0.16.x). See the [Authentication](./qwen-serve.md#authentication) section of the user guide for the canonical BYO setup.

The daemon reads the bearer token from either `--token <value>` on the CLI or the `QWEN_SERVER_TOKEN` env var (whitespace stripped from both). The TypeScript SDK's `DaemonClient` constructor falls back to `QWEN_SERVER_TOKEN` when no `token` option is passed (PR 27 fallback — clients with the env var set never need to thread the value through their script).

One shell-level `export` covers both server boot and SDK client construction.

## Linux: systemd user unit

`~/.config/systemd/user/qwen-serve.service`:

```ini
[Unit]
Description=Qwen Code daemon (loopback HTTP + SSE)
After=network.target

[Service]
Type=simple
# Replace with your project; %h expands to $HOME under user units.
WorkingDirectory=%h/your-project

> ExecStart=/usr/local/bin/qwen serve --hostname 127.0.0.1 --port 4170

Reviewer: test

Author: Resolved without code change — comment body is just "test", appears accidental.

> ExecStart=/usr/local/bin/qwen serve --hostname 127.0.0.1 --port 4170

Reviewer: **[Suggestion]** Both the systemd `ExecStart` (line 34) and launchd `<string>` (line 84) hardcode `/usr/local/bin/qwen` as the binary path, with no instruction to verify or adjust. This path is only correct for Homebrew on Intel Macs or a manual global install on some Linux distros. Most Node.js developers use nvm (`~/.nvm/versions/node/v22.x/bin/qwen`), fnm, volta, or Homebrew on Apple Silicon (`/opt/homebrew/bin/qwen`). A user copying the template verbatim gets an opaque "No such file or directory" from systemd/launchd.

Consider using a placeholder with a setup step:
```ini
# Run `which qwen` to find your binary path, then substitute:
ExecStart=/PATH/TO/qwen serve --hostname 127.0.0.1 --port 4170
```
Or add a one-liner before the template block:
```bash
QWEN_BIN=$(which qwen) && sed -i "s|/usr/local/bin/qwen|$QWEN_BIN|" ~/.config/systemd/user/qwen-serve.service
```
Apply the same treatment to the launchd plist `<string>` on line 84.

_— qwen3.7-max via Qwen Code /review_

Author: Adopted in 265891f23 — accurate catch. Both systemd `ExecStart` and launchd `ProgramArguments` now use `/PATH/TO/qwen` placeholder. Added a prominent "Find your `qwen` binary first" callout block above each template that lists the actual common locations (Linuxbrew, nvm, fnm, Volta on Linux; Apple Silicon Homebrew, Intel Homebrew, nvm, Volta on macOS) and points users at `which qwen`. Inline comments at the relevant lines reinforce "systemd / launchd does NOT read \$PATH". Should keep nvm/Volta/Apple-Silicon developers from hitting opaque "No such file or directory" on first start.
